### PR TITLE
Merge leo's contexts addon into the monorepo

### PR DESCRIPTION
**Pull request by ndelangen**
Issue: https://github.com/leoyli/addon-contexts

## What I did
I merged his repo into the monorepo

### CODE REVIEW

**Comment by ndelangen** [comment]
@leoyli Here's the merger, do you think you can help get it green and tested, and get some usage of it in the official example?

**Comment by shilman** [comment]
@ndelangen please, let's give this a label! :trollface:

File: addons/contexts/src/preview/api.ts [hunk 1]
```diff
@@ -30,8 +30,7 @@ export const addonContextsAPI = singleton(() => {
 
   // (Vue) hold a reference for updating props in its reactive system
   let reactivePropsMap = {};
-  const updateReactiveSystem = (propsMap: GenericObject) =>
-    Object.assign(reactivePropsMap, propsMap);
```
Comment (leoyli) [inline addons/contexts/src/preview/api.ts]: If I remembered correctly, this may break Vue reactive system.  I have to give it a look later of the day.

**Comment by leoyli** [comment]
Thank you all guys!  I just gave it a quick look now~

I will test as if it is working as expected when I back home...  Hope to merge it soon!

**Comment by leoyli** [comment]
@ndelangen,  I found found several typing and minor issues here.  I think it would be reasonable for me just append several commits after this PR.  Any concerns?!

**Comment by leoyli** [comment]
Also, in the current API setting, the parameter and setting object is not an `object` but an `array`...  Should I reshape it into `object`, or is it valid to accept an array?!

File: addons/contexts/src/preview/api.ts (hunk 1 above)
Comment (leoyli) [reply addons/contexts/src/preview/api.ts]: I have added a comment around this :)

**Comment by leoyli** [comment]
> @ndelangen, I have found several typing and minor issues here. I think it would be reasonable for me just append several commits on this PR. Any concerns?!

**Updated**: I've made several fixes to make the build working (ensure prettier, type-check, ts-lint) get green.  I have also manually tested the build version from the mono repo on my local machine: both React, and Vue are correctly functioned. The code is not push to this branch but a side branch (wait to get responded before I push it here or somewhere).

Depends on the scope of this PR, I see 2 related works: (1) add several stories to show case this addon in both official React and Vue; (2) add unit tests (<- this may take couple days..).

**Comment by ndelangen** [comment]
@leoyli Awesome work! I think adding usage to the official example is what's needed before merging.

**Comment by leoyli** [comment]
> @leoyli Awesome work! I think adding usage to the official example is what's needed before merging.

Okay!  I will add the example into both the official React and Vue later today into this PR.

**Comment by leoyli** [comment]
Voilà!  I think it is ready to be reviewed!

## What I did

- Fix typing, prettier, linting errors to make a green build.
- Add some useful code WHY comments for people interested in the source code.
- Add 2 stories under `Addon | Contexts` in our official Storybook (React).

---
@ndelangen, @shilman, @danielduan, @UsulPro : Please let me know if you have any feedback!

P.S. I was attempting to add new stories into **Vue kitchen sink**, but `dev` script throws me errors... any idea?!

**Comment by leoyli** [comment]
I have rebase the branch to resolve merge conflicts :)

**Comment by leoyli** [comment]
Voilà! I think it is ready to be reviewed!

## What I did
- Fix some bugs (storybook tab switches / story-level configs).
- Fix typing, prettier, linting errors, example story snapshots to get a green build.
- Add some WHY comments for people interested in the source code.
- Add 2 stories under `Addon | Contexts` in our official Storybook (React).

Please let me know if you have any feedback!

**Comment by leoyli** [comment]
Uh... it seems something mess up with my force rebasing. I will close it and reopen another PR.  Sorry about that!

**Pull request closed without merging**